src/hotspot/share/runtime/safepoint.cpp

```diff
@@ -976,17 +978,18 @@ void ThreadSafepointState::handle_polling_page_exception() {
       // to keep it in a handle.
       oop result = caller_fr.saved_oop_result(&map);
       assert(oopDesc::is_oop_or_null(result), "must be oop");
-      return_value = Handle(thread(), result);
+      return_value = Handle(self, result);
       assert(Universe::heap()->is_in_or_null(result), "must be heap pointer");
     }
 
     // We get here if compiled return polls found a reason to call into the VM.
     // One condition for that is that the top frame is not yet safe to use.
     // The following stack watermark barrier poll will catch such situations.
-    StackWatermarkSet::after_unwind(thread());
+    StackWatermarkSet::after_unwind(self);
 
     // Process pending operation
-    SafepointMechanism::process_if_requested(thread());
+    SafepointMechanism::process_if_requested(self);
+    self->check_and_handle_async_exceptions();
```

Comment:
I'm good with the addition of this call to check_and_handle_async_exceptions().

In particular because of this comment block in:

```
src/hotspot/share/runtime/thread.cpp:
void JavaThread::check_and_handle_async_exceptions(bool check_unsafe_error) {
  if (has_last_Java_frame() && has_async_condition()) {
    // If we are at a polling page safepoint (not a poll return)
    // then we must defer async exception because live registers
    // will be clobbered by the exception path. Poll return is
    // ok because the call we a returning from already collides
    // with exception handling registers and so there is no issue.
    // (The exception handling path kills call result registers but
    //  this is ok since the exception kills the result anyway).
```

so check_and_handle_async_exceptions() clearly was expecting to
be called when at a " poll return" which is the block that contains
this new call. That's all good!

What I can't quite figure out is how we "lost" the call to
check_and_handle_async_exceptions() that must have been
made somehow before. Clearly that function was expecting
to be called from a "poll return" site and that's what you're
putting back, but how was that call made before and what
change got rid of it?

Reply:
The same problematic execution sequence detailed in 8223572 for the external_suspend case was there for async exceptions too. So, while the thread was blocked in the handshake a ThreadStop() operation could have been executed. After returning from the handshake there was no async exception check. We probably never noticed it because the throw would happen anyways on some later transition back to java (vm->java or native->java). 
As to how was that call made before and what change got rid of it: this code was there before handshakes existed. ThreadStop() was implemented as a safepoint and we were probably relying on the special_runtime_exit_condition() check at the end of SS::block() (exactly like the one I'm removing from ~TIVFH and which I'll try to get rid of to eliminate recursions).

Reply:
Makes sense. Thanks!

Reply:
Not to belabour this but I don't understand what this sentence:

// If we are at a polling page safepoint (not a poll return)

is actually making a distinction between. What is a "poll return" here versus a "polling page safepoint"?

Reply:
"poll return" would be a poll immediately before a return (also found out by reading it from the comments in ThreadSafepointState::handle_polling_page_exception()). So unless the poll in the nmethod was done right before a return, we defer the async exception. There is also a comment before the has_special_runtime_exit_condition() check on SS::block() that explains the reason:

```
  // Note: we never deliver an async exception at a polling point as the
  // compiler may not have an exception handler for it. The polling
  // code will notice the async and deoptimize and the exception will
  // be delivered. (Polling at a return point is ok though). Sure is
  // a lot of bother for a deprecated feature...
```